### PR TITLE
*: optimize [int column] <cmp> [constant] (#3911)

### PR DESCRIPTION
**Pull request by blacktear23**
Let compareFunction can process string cast to Int or Decimal not just throw a Real as default.
#3911

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
You may change the behavior of getCmpType like the following description when 
meeting `[int column] <cmp> [non-int constant]`.
``` sql
create table t(a int)
```

-  \<cmp\> = *LT*
`a < 1.1` --> `a <= 1`
- \<cmp\> = *LE*
`a <= 1.1` --> `a <= 1`
- \<cmp\> = *GT*
`a > 1.1` --> `a >= 2`
- \<cmp\> = *GE*
`a >= 1.1` --> `a >= 2`
- \<cmp\> = *NE*
`a != 1.1` --> `a >= 2 || a <= 1`
- \<cmp\> = *EQ*
`a == 1.1` --> `1 == 0`
- \<cmp\> = *NullEQ*
`a <=> 1.1` --> `a >= 2 || a <= 1`

you may add `ceiling` or `floor` above the constant arg, and `WrapWithCastAsInt` to implement the above behavior.

**Comment by blacktear23** [comment]
@XuHuaiyu in my opinion, make cast(Column) operator down to TiKV will give DBA more clear SQL plan. Although this Int compareTo Float optimization is correct, but change SQL's AST would be a bad idea.

**Comment by XuHuaiyu** [comment]
yeah, you are right, we are working on supporting cast in TiKV.

Here we only optimize the compare funcs when meeting <int column> cmp <non-int column>
this can provide us a benefit when calculating the key range.
Otherwise, we only change the expression of one ast node, this will not influence the ast tree, and `FoldConstant` will be called after rewriting which can avoid a complicated expression and not confuse the reader.

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -268,6 +269,44 @@ func getCmpType(a types.TypeClass, b types.TypeClass) types.TypeClass {
...
+			return types.ClassDecimal
+		}
+		if erri != nil && errf != nil {
+			return t
+		}
+	}
+	return pt
+}
+
+// Try to fix Issue3911
```
Comment (winoros) [inline expression/builtin_compare.go]: This comment is not suitable at the head of one function.

**Comment by blacktear23** [comment]
@XuHuaiyu can any body give me a help to implement the optimization as you mentioned?

**Comment by XuHuaiyu** [comment]
@blacktear23 yeah, if you get any problem, you can submit a comment here.
we'll reply as soon as possible.

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -268,6 +270,75 @@ func getCmpType(a types.TypeClass, b types.TypeClass) types.TypeClass {
 	return types.ClassReal
 }
 
+// smartConvertType will check expression's value and compare to primary type (tp)
```
Comment (jackysp) [inline expression/builtin_compare.go]: primary type (tp) should be (pt)?

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -281,6 +352,87 @@ func isTemporalColumn(expr Expression) bool {
 	return true
 }
 
+// tpInt
```
Comment (jackysp) [inline expression/builtin_compare.go]: what is meaning of tpInt here?

File: expression/builtin_compare.go [hunk 4]
```diff
@@ -268,6 +270,75 @@ func getCmpType(a types.TypeClass, b types.TypeClass) types.TypeClass {
...
+//    a > "1.0"     => ClassInt
+//    a > "1.1"     => ClassDecimal
+//    a > "notNum"  => ClassString
+//    a > 1.1       => ClassDecimal
+//    a > 1.0       => ClassInt
+// a is ClassDecimal Column:
+//    a > "1"       => ClassDecimal
+//    a > "1.1"     => ClassDecimal
+//    a > "notNum"  => ClassString
+func smartConvertType(expr Expression, pt, t types.TypeClass) types.TypeClass {
```
Comment (jackysp) [inline expression/builtin_compare.go]: Please add some tests.

**Comment by XuHuaiyu** [comment]
I guess I didn't make myself clear,
1. We *do not* need to concern about whether the constant is *a decimal* or not, we **just** concern about whether the compare expression is *\<int column\> cmp \<consant\>*, we can make this check before `getCmpType`.
2. For not change the `funcName`, and make the implementation more easier, just follow the following rules:
-  \<cmp\> = *LT*
`a < 1.1` --> `a < ceil(1.1)`
`1.1 < a` --> `floor(1.1) < a`
- \<cmp\> = *GE*
`a >= 1.1` --> `a >= ceil(1.1)`
`1.1 >= a` --> `floor(1.1) >= a`
- \<cmp\> = *LE*
`a <= 1.1` --> `a <= floor(1)`
`1.1 <= a` --> `ceil(1.1) <= a`
- \<cmp\> = *GT*
`a > 1.1` --> `a > floor(1.1)`
`1.1 > a` --> `ceil(1.1) > a`
3. We can define functions like 
``` go
func handleIntColCmpConst(arg0 Expression, arg1 Expresson, cmpOp, ctx) (bool, Expression, Expression, error){
    if !((arg0.isIntColumn && arg1.isConstant) ||  (arg1.isIntColumn && arg0.isConstant)){
        return false, ...
    }
   switch cmpOp{
     constant = wrapWithCeilOrFloor(constant, ast.Ceil/ ast.Floor, ctx)
   }
   return true, column, constant, nil 
}
func(c *compareFunctionClass) wrapWithCeilOrFloor(arg Expression, funcName, ctx) (Expression, error){// funcName ∈[ast.Ceil, ast.Floor]
    return NewFunction(ctx, funcName, types.NewFieldType(mysql.TypeLonglong), arg)
}

// in getFunction() {}
we check
if handleIntColCmpConst == true{
    sig, err = c.generateCmpSigs(args, tpInt, ctx)
    return ...
}
```

**Comment by blacktear23** [comment]
@XuHuaiyu gotcha, And I think I should refactor and move logic to scalar_function.go. And I also find out how to handle EQ operator, but NE operator is much more difficulty.

File: expression/builtin_compare.go (hunk 3 above)
Comment (blacktear23) [inline expression/builtin_compare.go]: It's just a hint when I write this function, I will refactor all those. ;)

**Comment by blacktear23** [comment]
@XuHuaiyu this implementian not wrap Ceil or Floor on Constant It do convert in optimize function. For ast.Ceil or ast.Floor will those op pushed down to TiKV?

**Comment by blacktear23** [comment]
After I execute `explain select * from indextest where success > floor("1.1");` I found the plan like below:

```
*************************** 1. row ***************************
      ID: TableScan_4
    Json: {
    "db": "test",
    "table": "indextest",
    "desc": false,
    "keep order": false,
    "push down info": {
        "limit": 0,
        "access conditions": null,
        "index filter conditions": null,
        "table filter conditions": null
    }
}
ParentID: Selection_2
*************************** 2. row ***************************
      ID: Selection_2
    Json: {
    "condition": [
        "gt(cast(test.indextest.success), 1)"
    ],
    "scanController": false,
    "child": "TableScan_4"
}
ParentID:
```

That means if I wrap a ast.CEIL function to Constant argument will cause a Int column compare to a Decimal constant, and then TiDB will do a full table scan on TiKV and gather all data to TiDB to do selection.

**Comment by hanfei1991** [comment]
duplicate with #4286

**Pull request closed without merging**